server/db/call_store.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	_, err = s.wDB.Exec(q, args...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run query: %w", err)
+	}
+
+	return call, nil
+}
+
+func (s *Store) UpdateCall(call *public.Call) error {
+	s.metrics.IncStoreOp("UpdateCall")
+
+	if call == nil {
+		return fmt.Errorf("call should not be nil")
+	}
+
+	qb := getQueryBuilder(s.driverName).
+		Update("calls").
+		Set("EndAt", call.EndAt).
+		Set("DeleteAt", call.DeleteAt).
+		Set("Participants", call.Participants).
+		Set("Stats", call.Stats).
+		Set("Props", call.Props).
+		Where(
+			sq.Eq{"ID": call.ID},
+			sq.Eq{"ChannelID": call.ChannelID},
+		)
+
+	q, args, err := qb.ToSql()
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	res, err := s.wDB.Exec(q, args...)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %w", err)
+	}
+
+	count, err := res.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed to get rows affected: %w", err)
+	}
+
+	if count != 1 {
+		return fmt.Errorf("failed to update call")
+	}
+
+	return nil
+}
+
+func (s *Store) DeleteCall(callID string) error {
+	s.metrics.IncStoreOp("DeleteCall")
+
+	qb := getQueryBuilder(s.driverName).
+		Update("calls").
+		Set("DeleteAt", time.Now().UnixMilli()).
+		Where(sq.Eq{"ID": callID})
+
+	q, args, err := qb.ToSql()
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	res, err := s.wDB.Exec(q, args...)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %w", err)
+	}
+
+	count, err := res.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed to get rows affected: %w", err)
+	}
+
+	if count != 1 {
+		return fmt.Errorf("failed to delete call")
+	}
+
+	return nil
+}
+
+func (s *Store) DeleteCallByChannelID(channelID string) error {
+	s.metrics.IncStoreOp("DeleteCallByChannelID")
+
+	qb := getQueryBuilder(s.driverName).
+		Update("calls").
+		Set("DeleteAt", time.Now().UnixMilli()).
+		Where(sq.Eq{"ChannelID": channelID})
+
+	q, args, err := qb.ToSql()
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	_, err = s.wDB.Exec(q, args...)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Store) GetCall(callID string, opts GetCallOpts) (*public.Call, error) {
+	s.metrics.IncStoreOp("GetCall")
+
+	qb := getQueryBuilder(s.driverName).Select("*").
+		From("calls").
+		Where(sq.Eq{"ID": callID})
+
+	q, args, err := qb.ToSql()
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	var call public.Call
+	if err := s.dbXFromGetOpts(opts).Get(&call, q, args...); err == sql.ErrNoRows {
+		return nil, fmt.Errorf("call not found")
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to get call: %w", err)
+	}
+
+	return &call, nil
+}
+
+func (s *Store) GetCallByChannelID(channelID string, opts GetCallOpts) (*public.Call, error) {
```

Comment:
There's a bit of redundancy here given the `CallID` vs `ChannelID` conundrum. We could combine by doing a simple `OR` but for now I am keeping them separate as we may not need them both, depending on how we implement things.

Reply:
Sure, difference is simple though. `CallID` is unique, hence why it will be primary key here, `ChannelID` on the other hand is shared by all of the calls that will ever happen in a channel. However, since we don't support multiple concurrent calls in the same channel, `channelID` is what's usually used by clients and APIs to identify the current (active) call.